File: 백준/Silver/14225. 부분수열의 합/부분수열의 합.js
const fs = require('fs')

const read =fs.readFileSync('dev/stdin').toString().trim();

let [n, arr] = read.split("\n");
n = Number(n)
arr = arr.split(" ").map(Number)

const main = () => {
    const mask = new Array(100000 * 20 + 1).fill(false);
    const li = [];
    const recur = (sumv) => {
        if (li.length) mask[sumv] = true;
        if (li.length == n) return;
        
        for(let i = 0 ; i < arr.length; i++ ){
            if(!li.length || i > li.at(-1)) {
                li.push(i)
                recur(sumv + arr[i]);
                li.pop()
            }
        }
    }
    
    recur(0);
    
    let i = 1
    while (mask[i]) i++;
    return i;
}

console.log(main())

File: 백준/Silver/14225. 부분수열의 합/부분수열의 합.py
import sys

input = sys.stdin.readline

n = int(input())
arr = [int(i) for i in input().split()]

def main():
    li = []
    mask = [False] * (100000 * 20 + 1)
    def recur(sumv):
        if li:
            mask[sumv] = True
        if len(li) == n:
            return
        
        for i in range(len(arr)):
            if not li or i > li[-1]:
                li.append(i)
                recur(sumv + arr[i])
                li.pop()
    recur(0)
    i = 1
    while mask[i]:
        i += 1
        
    return i

print(main())